subsys/nrf_security/src/drivers/cracen/cracen_sw/src/cracen_sw_aes_ctr.c

```diff
@@ -0,0 +1,235 @@
+#define AES_BLOCK_LAST_BYTE_INDEX  (SX_BLKCIPHER_AES_BLK_SZ - 1)
+#define AES_CTR_COUNTER_START_BYTE (SX_BLKCIPHER_AES_BLK_SZ - 4)
+
+/* Increments the counter by one block */
+static void increment_counter(uint8_t *ctr)
+{
+	/* CTR mode increments the rightmost 32 bits in big-endian */
+	for (int i = AES_BLOCK_LAST_BYTE_INDEX; i >= AES_CTR_COUNTER_START_BYTE; i--) {
+		if (++ctr[i] != 0) {
+			break;
+		}
+	}
+}
+
+psa_status_t cracen_sw_aes_ctr_setup(cracen_cipher_operation_t *operation,
+				     const psa_key_attributes_t *attributes,
+				     const uint8_t *key_buffer, size_t key_buffer_size)
+{
+	psa_status_t status;
+
+	/* Validate AES key size */
+	if (key_buffer_size != 16 && key_buffer_size != 24 && key_buffer_size != 32) {
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	/* Verify key buffer fits in operation structure */
+	if (key_buffer_size > sizeof(operation->key_buffer)) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	/* Copy the key into the operation struct */
+	memcpy(operation->key_buffer, key_buffer, key_buffer_size);
+
+	/* Load the key reference for ECB operations */
+	status = cracen_load_keyref(attributes, operation->key_buffer, key_buffer_size,
+				    &operation->keyref);
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	/* Initialize the operation */
+	operation->alg = PSA_ALG_CTR;
+	operation->dir = CRACEN_ENCRYPT; /* CTR mode is identical for encrypt/decrypt */
+	operation->blk_size = SX_BLKCIPHER_AES_BLK_SZ;
+	operation->unprocessed_input_bytes = 0;
+	/* We don't consider the initalization finalized until IV is set */
+	operation->initialized = false;
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t cracen_sw_aes_ctr_set_iv(cracen_cipher_operation_t *operation, const uint8_t *iv,
+				      size_t iv_length)
+{
+	if (operation->alg != PSA_ALG_CTR) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (iv_length != SX_BLKCIPHER_AES_BLK_SZ) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+	operation->unprocessed_input_bytes = 0;
+	/* Copy the IV (which includes the initial counter) */
+	memcpy(operation->iv, iv, iv_length);
+	operation->initialized = true;
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t cracen_sw_aes_ctr_update(cracen_cipher_operation_t *operation, const uint8_t *input,
+				      size_t input_length, uint8_t *output, size_t output_size,
+				      size_t *output_length)
+{
+	psa_status_t status = PSA_SUCCESS;
+	size_t bytes_written = 0;
+	size_t remaining_input;
+	const uint8_t *current_input;
+	uint8_t *current_output;
+	uint8_t *keystream_block;
+	size_t keystream_used;
+	size_t remaining_keystream_bytes;
+	size_t bytes_to_process;
+
+	*output_length = 0;
+
+	if (!operation->initialized) {
+		return PSA_ERROR_BAD_STATE;
+	}
+
+	/* Valid operation, we just don't do anything */
+	if (input_length == 0) {
+		return PSA_SUCCESS;
+	}
+
+	/* CTR is a stream mode operation so output must be same length as input */
+	if (output == NULL || output_size < input_length) {
```

Comment:
This is not in line with `psa_cipher_update` that returns `PSA_ERROR_INVALID_ARGUMENT` when buffer pointers are null.

Reply:
You are correct. That should be output_size, not output.